### ISSUE TITLE
Roles & permissions

### ISSUE THREAD

**Issue by taoeffect** [okTurtles/group-income#202]
### Problem

Some users will likely need to perform special functions for the group. It's likely that some of these functions should be restricted only to those users with the assigned role to perform them.

### Solution

Design a generic role + permissions system, where each user can be assigned a role that comes with certain permissions.

A "role" is a bundle of permissions.

A permission is a string that is internally interpreted by both the UI and the contracts as being allowed to perform certain actions.

Possibly useful stuff:

- https://github.com/leonardovilarinho/vue-acl

**Comment by Silver-IT** [comment]
@taoeffect, @leihla. Seems like we need design for this.

**Comment by SebinSong** [comment]
## Development Steps

**Figma link for the hypothetical UIs:** [here](https://www.figma.com/design/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?node-id=27350-54034&t=1gasEw69Na8Z0lET-0)

### Step 1. - Build 'Roles & Permissions' UI with dummy data ✅ 
Building **'Roles & Permissions'** section in **'Group Settings'** page, which will look something like below.
(NOTE: below is just an example design and the actual UI might look different from it.)

<img src='https://github.com/user-attachments/assets/b1d39393-3524-48b3-b320-9c1591ef3229' width='480'>

### Step 2. - Build 'Add Permissions' modal / Implement `group.js` action for it.  ✅ 
- Build 'Add Permissions' modal UIs
- Create a group contract action for adding permissions for a group member
- Display actual roles & permission details of group members in `GroupSettings.vue` (This table will only be exposed for a member who has been granted `view permissions` permission item.)

<img width="530" src="https://github.com/user-attachments/assets/e4791c25-b730-4d26-be27-724a2ee19d70" />

### Step 3. Build UIs / Implement 'Update Permissions' and 'Remove Permissions' ✅ 

- Build UIs / Implement front-end logics for 'Update Permissions' functionality

<img width="530"  src="https://github.com/user-attachments/assets/ada71330-3612-4a8b-8408-8cb0b1206215" />

- Build UIs / Implement front-end logics for 'Remove Permissions' functionality

<img width="530" src="https://github.com/user-attachments/assets/8411ca56-e3d9-4994-8f12-fad2dcfc8440" />

### Step 4. Conditionally enable features / functionalities based on permissions granted
- Eg. `Delete chatroom` functionality is only possible via members with `delete-channel` permission etc.

### Step 5. Write Cypress test suites for roles & permissions
 
---
cc. @taoeffect

**Comment by SebinSong** [comment]
Hi @taoeffect ,

I'm in the middle of defining some roles in a group now, and got a couple of questions regarding the role `moderator - delegator` we discussed last week.

**Q1.**
The English phrase **'delegate permissions'** doesn't make 100% sense to me yet, tbh. Does it means the ability to add/edit/remove permissions for a group member on behalf of the group administrator?

**Q2.**
If the answer to Q1. is yes, does it make **no** sense to have both `delegate-permissions` and `add/edit/remove permissions` in permission items of `moderator - delegator` like below screenshot? Because the phrase 'delegate permissions' means the ability to perform those three actions?

<img src='https://github.com/user-attachments/assets/2471bf2d-49c1-4669-8734-4f7a070b4e3a' width='380'>

**Q3.**
Another question is about the difference between `moderator - delegator` and `moderator`. We won't allow moderator to perform `add/edit/remove permissions` for a group member but only give them a permission to `view` permissions set by the admin. is it correct?

Thanks,

**Comment by taoeffect** [comment]
> The English phrase 'delegate permissions' doesn't make 100% sense to me yet, tbh. Does it means the ability to add/edit/remove permissions for a group member on behalf of the group administrator?

So, the idea I had was that if you were a "delegator", it meant that the admin had trusted you to be able to give whatever permissions you have to other people. It would mean you (as the admin) trust them to decide who else can have these permissions, and it would necessarily imply the ability to remove permissions too.

However, it would not mean that the delegator could delegate the *delegate-permission* to others. Only the admin would be able to do this. In other words, if the admin granted me `moderator (delegator)` role, I would only be able to grant `moderator` permissions to others, but not `moderator (delegator)`. This is something I didn't mention on the call.

This is to prevent the following scenario:

1. Admin grants `moderator-delegator` to A
2. A grants `moderator-delegator` to B
3. Admin leaves the group or goes on vacation for a while
4. B removes A's permissions

In that scenario, the moderator-delegator A made a mistake in selecting B as another moderator-delegator and the group gets overthrown. I have personally witnessed this scenario happen.

In other words, a `moderator (delegator)` cannot create another `moderator (delegator)`, only a `moderator`. Only the `admin` can create delegators.

So, I don't think `add/edit/remove permissions` is necessary here. Just `delegate-permissions` — which if granted, allows that person to delegate all permissions they have except for `delegate-permissions`, as well as remove `moderator` permissions from others.

Hopefully that answers it, but if anything is unclear let me know your doubts/concerns on Slack. :)

**Comment by SebinSong** [comment]
Hi @taoeffect 

I'm looking at [TODO step 4](https://github.com/okTurtles/group-income/issues/202#issuecomment-2567190831) of this issue which is 'enable/disable a certain feature based on whether the permission is given or not'.

There are currently 6 permissions I've added as part of step 1,2 & 3(currently active PR [here](https://github.com/okTurtles/group-income/pull/2895) FYI) and they were taken from the prototype drawing you gave me during the call I had with you last year regaring this task:

<img width="540" src="https://github.com/user-attachments/assets/416cc173-4268-4584-b71b-a51254c5bbbe" />

but you might have given some of these as just placeholders, so need your confirmation before I proceed with further development. Could you confirm if below are correct?

1) **View permission** : ✅ 
By default, the user can only see their own permission. If this permission is given, they can see other group memeber's permissions too.

2) **Assign delegator**: ✅ 
It is a permission to assign 'moderator - delegator' role to any group member and is exclusively given to the group admin per [this comment](https://github.com/okTurtles/group-income/issues/202#issuecomment-2574267136).

3) **Delegate permissions** ✅ 
Permission to delegate(add/edit/remove, correc me if I'm wrong though) permissions to others.
This is only given to `group admin` and `moderator - delegator` roles.

4) **Remove member**  ✅ 
A group member can be kicked out by voting on proposal but members with this permission can remove a member without creating a proposal. Currently this is only enabled for the group creator.

5) **Revoke invite** ✅
Any invite created via `add-member` proposal can only be viewed/revoked by the owner(The member who opened the proposal). A member with this permission has a right to revoke these invites regardless of whether they opened the proposal or not.

6) **Delete channel** 🛠️
Currently a chatroom(whether private or public) can only be deleted by someone who created it. A member with this permission can delete any channel they are part of regardless of whether they are creators or not.

---

Pls let me know if any of these are not correct or if there is any other permissions we would need to add.
(Tagging @dotmacro here as well, in case you need discussion with her on any of these)
 
Thanks,

**Comment by taoeffect** [comment]
Yeah that looks about right!
